dotCMS/src/main/java/com/dotcms/rest/MapToContentletPopulator.java

```diff
@@ -383,53 +434,158 @@ List<Category> categories(final Contentlet contentlet, final User user, final bo
+     * @param contentlet
+     * @param user
+     * @return
+     * @throws DotDataException
+     * @throws DotSecurityException
+     */
+    Map<CategoryField, Set<Category>> fetchExistingCategories(final Contentlet contentlet, final User user)
+            throws DotDataException, DotSecurityException {
+
+        final ImmutableMap.Builder<CategoryField, Set<Category>> builder = ImmutableMap.builder();
+        //Since we can not rely there will always be a inode on the incoming request  we must lookup the current working contentlet
+        final Contentlet workingContentlet = APILocator.getContentletAPI()
+                .findContentletByIdentifier(contentlet.getIdentifier(), false,
+                        contentlet.getLanguageId(), APILocator.systemUser(), false);
+        if(null != workingContentlet){
+            final CategoryAPI categoryAPI = APILocator.getCategoryAPI();
+            //Parents are the categories assigned to the contentlet. But they this does not tell you how they are assigned through the fields
+            final List<Category> parents = categoryAPI.getParents(workingContentlet, user, false);
+            final List<CategoryField> categoryFields = workingContentlet.getContentType()
+                    .fields(CategoryField.class).stream().map(field -> (CategoryField) field)
+                    .collect(Collectors.toList());
+            for (final CategoryField categoryField : categoryFields) {
+                //We need to know how categories are saved by field on the given contentlet
+                final Set<Category> selected = findSelected(categoryField, parents, user);
+                if(null != selected){
+                   builder.put(categoryField, selected);
+                }
             }
         }
+        return builder.build();
+    }
+
+    /**
+     * This is basically tells how we bind categories to the fields they belong to
+     * @param categoryField
+     * @param categories
+     * @param user
+     * @return
+     * @throws DotDataException
+     * @throws DotSecurityException
+     */
+    Set<Category> findSelected(final CategoryField categoryField, final List<Category> categories,
+            final User user)
+            throws DotDataException, DotSecurityException {
+        final CategoryAPI categoryAPI = APILocator.getCategoryAPI();
+        final Category parent = categoryAPI.find(categoryField.values(), user,
+                false);
+
+        if (categoryAPI.canUseCategory(parent, user, false)) {
+            return categories.stream()
+                    //in order to make sure we include all categories We need to take into account also the actual parent and verify it against the current category
+                    .filter(category -> category.equals(parent) || categoryAPI.isParent(category, parent, user, false))
+                    .collect(Collectors.toSet());
+        }
+        return null;
```

Comment:
and empty set could be better?

Reply:
I made it return null to be able to differentiate the case. Actually the statement that filters can be empy 
I making a validation to ignore when null